### ISSUE TITLE
Unexpected diagnostic on super(..)

### ISSUE THREAD

**Issue by JeroenBos** [microsoft/pylance-release#589]
## Environment data

-   Language Server version: ms-python.vscode-pylance 2020.11.0
-   OS and version: Ubuntu 18.04.3
-   Python version (& distribution if applicable, e.g. Anaconda): v3.8.2, ms-python.python 2020.10.332292344

## Expected behaviour

No diagnostic to be reported.

## Actual behaviour

The following diagnostic is reported on the code below
```
Expected class type as first argument to "super" call but received "(*args: Any, **kwargs: Any) -> APITests"
```

## Code Snippet / Additional information

```python
from unittest import mock


@mock.patch("lib.authorization._get_token_auth_header")
class APITests:
    def setUp(self):
        super(APITests, self).setUp()
```

Weirdly enough, when I remove the line `@mock.patch` the diagnostics is no longer reported.

I use the pattern `super(<class_name>, self).m()` all over the place, but only with certain decorators will this result in a diagnostics (for instance for `@unittest.skipUnless` it does _not_ report a diagnostic). 

In version 2020.10.0 the diagnostic is not reported, but since 2020.10.1 it is.

Some screenshots for convenience and clearer communication:

![image](https://user-images.githubusercontent.com/15310631/98696717-6e619700-2374-11eb-9483-33ed2393c062.png)

![image](https://user-images.githubusercontent.com/15310631/98696690-64d82f00-2374-11eb-883a-babac9cd3759.png)

## Logs
When I restart vscode and only open that file, trace logs show the following:
<details>

```
[Info  - 4:39:28 PM] Pylance language server 2020.11.0 (pyright 9d8e431b) starting
[Info  - 4:39:28 PM] Server root directory: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist
[Info  - 4:39:28 PM] No configuration file found.
[Info  - 4:39:28 PM] Setting pythonPath for service "remin-api": "/home/jeroen/git/remin-api/venv/bin/python"
Search paths found for configured python interpreter:
  /usr/local/lib/python3.8
  /usr/local/lib/python3.8/lib-dynload
  /home/jeroen/git/remin-api/venv/lib/python3.8/site-packages
[Error - 4:39:28 PM] stubPath /home/jeroen/git/remin-api/typings is not a valid directory.
[Info  - 4:39:28 PM] Assuming Python version 3.8
[Info  - 4:39:28 PM] Assuming Python platform Linux
[Info  - 4:39:28 PM] Searching for source files
[Info  - 4:39:29 PM] Auto-excluding /home/jeroen/git/remin-api/venv
[Info  - 4:39:29 PM] Found 97 source files
[Info  - 4:39:29 PM] Background analysis(1) root directory: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist
[Info  - 4:39:29 PM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: getSemanticTokens
[BG(1)] parsing: /home/jeroen/git/remin-api/test/resources/error.py (18ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/2and3/builtins.pyi [fs read 3ms] (129ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/2and3/builtins.pyi (59ms)
[BG(1)] binding: /home/jeroen/git/remin-api/test/resources/error.py (1ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/__init__.pyi [fs read 1ms] (19ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/__init__.pyi ...
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/async_case.pyi [fs read 0ms] (1ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/async_case.pyi (1ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/case.pyi [fs read 2ms] (34ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/case.pyi (24ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/loader.pyi [fs read 2ms] (4ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/loader.pyi (1ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/main.pyi [fs read 0ms] (4ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/main.pyi (2ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/runner.pyi [fs read 0ms] (3ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/runner.pyi (1ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/signals.pyi [fs read 1ms] (1ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/signals.pyi (1ms)
[BG(1)]   parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/suite.pyi [fs read 0ms] (0ms)
[BG(1)]   binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/suite.pyi (1ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/__init__.pyi (83ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi [fs read 0ms] (12ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (1ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi [fs read 0ms] (2ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (1ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/typing.pyi [fs read 0ms] (38ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/typing.pyi (14ms)
[BG(1)] parsing: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/mock.pyi [fs read 1ms] (14ms)
[BG(1)] binding: /home/jeroen/.vscode/extensions/ms-python.vscode-pylance-2020.11.0/dist/typeshed-fallback/stdlib/3/unittest/mock.pyi (5ms)
Background analysis message: getSemanticTokens
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: /home/jeroen/git/remin-api/test/resources/error.py ...
[BG(1)]   checking: /home/jeroen/git/remin-api/test/resources/error.py (3ms)
[BG(1)] analyzing: /home/jeroen/git/remin-api/test/resources/error.py (3ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

</details>

**Comment by erictraut** [comment]
Thanks for the bug report.

The problem is that you are using a decorator that replaces the class `APITests` with an object that has a `__callable__` method that allocates an `APITests` instance. This will run without crashing, but it is highly unusual and prevents the static analysis of the `super` call from understanding the target of the `super` call.

You can work around this by adding the following cast:
```python
cast(Any, APITests)
```

I don't think we should eliminate this diagnostic because it is generally useful, but there's an argument to be made that it should be a "general type error" so it is suppressed when typeCheckingMode is set to "off". I've made that change, and it will be included in the next release.

**Comment by JeroenBos** [comment]
Okay, thank you. I didn't expect that the analyzer would draw such conclusions from decorator implementations, and that is a great feature. I also acknowledge the unusualness of this pattern. 

However, given that `unittest.mock.patch` is used very frequently, I wonder if the analyzer is in the business of special casing "well-known" symbols (like this decorator). If so I would propose to add this to such a list.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2020.11.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/master/CHANGELOG.md#2020111-11-november-2020
